api/src/main/java/me/lokka30/treasury/api/permission/context/ContextSet.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * This file is/was part of Treasury. To read more information about Treasury such as its licensing, see <https://github.com/ArcanePlugins/Treasury>.
+ */
+
+package me.lokka30.treasury.api.permission.context;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import org.jetbrains.annotations.Contract;
+import org.jetbrains.annotations.NotNull;
+
+public final class ContextSet {
+
+    private static final ContextSet EMPTY = new ContextSet(Collections.emptyMap());
+
+    @NotNull
+    public static ContextSet empty() {
+        return EMPTY;
+    }
+
+    @NotNull
+    public static ContextSet of(@NotNull Context @NotNull ... contexts) {
+        Objects.requireNonNull(contexts, "contexts");
+        if (contexts.length == 0) {
+            return empty();
+        }
+        Map<String, Context> map = new ConcurrentHashMap<>();
+        for (Context context : contexts) {
+            map.put(context.getKey(), context);
+        }
+        return new ContextSet(map);
+    }
+
+    private Map<String, Context> contexts;
+
+    @Contract(pure = true)
+    private ContextSet(Map<String, Context> contexts) {
+        this.contexts = contexts;
+    }
+
+    @NotNull
+    public Optional<Context> getContext(@NotNull String contextKey) {
+        Objects.requireNonNull(contextKey, "contextKey");
+        return Optional.ofNullable(this.contexts.get(contextKey));
+    }
+
+    public void registerContext(@NotNull Context context) {
```

Comment:
This method seems out-of-place. I thought ContextSet was immutable, hence `ContextSet.empty()` and your use of Collections.emptyMap(), which is immutable.

It's best to decide whether you want ContextSet to be mutable and maintain consistency in your decision afterward.

Reply:
Yeah I also think I shall transform this immutable.